wtxmgr/tx_test.go

```diff
@@ -1364,3 +1364,217 @@ func TestRemoveUnminedTx(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We'll insert both of them into the store without confirming them.
+	if err := store.InsertTx(ns, firstSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, secondSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure that both spends are found within the unconfirmed transactions
+	// in the wallet's store.
+	unminedTxs, err := store.UnminedTxs(ns)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(unminedTxs) != 2 {
+		t.Fatalf("expected 2 unmined txs, got %v", len(unminedTxs))
+	}
+
+	// Then, we'll confirm either the first or second spend, depending on
+	// the boolean passed, with a height deep enough that allows us to
+	// succesfully spend the coinbase output.
+	coinbaseMaturity := int32(chaincfg.TestNet3Params.CoinbaseMaturity)
+	bMaturity := BlockMeta{
+		Block: Block{Height: b100.Height + coinbaseMaturity},
+		Time:  time.Now(),
+	}
+
+	var confirmedSpendRec *wtxmgr.TxRecord
+	if first {
+		confirmedSpendRec = firstSpendRec
+	} else {
+		confirmedSpendRec = secondSpendRec
+	}
+	if err := store.InsertTx(ns, confirmedSpendRec, &bMaturity); err != nil {
+		t.Fatal(err)
+	}
+
+	// This should now trigger the store to remove any other pending double
+	// spends for this coinbase output, as we've already seen the confirmed
+	// one. Therefore, we shouldn't see any other unconfirmed transactions
+	// within it.
+	unminedTxs, err = store.UnminedTxs(ns)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(unminedTxs) != 0 {
+		t.Fatalf("expected 0 unmined txs, got %v", len(unminedTxs))
+	}
+}
+
+// TestInsertMempoolDoubleSpendConfirmedFirstTx ensures that when a double spend
+// occurs and both spending transactions are present in the mempool, if the
+// first spend seen is confirmed, then the second spend transaction within the
+// mempool should be removed from the wallet's store.
+func TestInsertMempoolDoubleSpendConfirmedFirstTx(t *testing.T) {
+	t.Parallel()
+	testInsertMempoolDoubleSpendTx(t, true)
+}
+
+// TestInsertMempoolDoubleSpendConfirmedFirstTx ensures that when a double spend
+// occurs and both spending transactions are present in the mempool, if the
+// second spend seen is confirmed, then the first spend transaction within the
+// mempool should be removed from the wallet's store.
+func TestInsertMempoolDoubleSpendConfirmSecondTx(t *testing.T) {
+	t.Parallel()
+	testInsertMempoolDoubleSpendTx(t, false)
+}
+
+// TestInsertConfirmedDoubleSpendTx tests that when one or more double spends
+// occur and a spending transaction confirms that was not known to the wallet,
+// then the unconfirmed double spends within the mempool should be removed from
+// the wallet's store.
+func TestInsertConfirmedDoubleSpendTx(t *testing.T) {
+	t.Parallel()
+
+	store, db, teardown, err := testStore()
+	defer teardown()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dbtx, err := db.BeginReadWriteTx()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer dbtx.Commit()
```

Comment:
I think we should really commit the db transaction between insertions/extractions, as this will be what happens in real use.

Reply:
Fixed.